src/srun/libsrun/fname.c

```diff
@@ -217,6 +218,19 @@ extern fname_t *fname_create(srun_job_t *job, char *format, int task_count)
 				break;
 			case 'x':
 				xmemcat(name, q, p - 1);
+                jobname = getenvp(job->env, "SLURM_JOB_NAME");
+                if (jobname != NULL) {
+                    jobnamecopy = strdup(jobname);
+                    xstrfmtcat(name, "%s", basename(jobnamecopy));
+                    xfree(jobnamecopy);
+                    jobnamecopy = NULL;
+                }
+				q = ++p;
+                xfree(jobname);
```

Comment:
Hmm, no? You get a pointer into the environment, please do not free that :)

Reply:
This will either abort (since the env lacks the magic cookie) or reach free(p). Neither is what you want, I think.

~~~~
void slurm_xfree(void **item, const char *file, int line, const char *func)
{
	if (*item != NULL) {
		size_t *p = (size_t *)*item - 2;
		/* magic cookie still there? */
		xmalloc_assert(p[0] == XMALLOC_MAGIC);
		p[0] = 0;	/* make sure xfree isn't called twice */
		free(p);
		*item = NULL;
	}
}
~~~~

~~~~
#if NDEBUG
#  define xmalloc_assert(expr)  ((void) (0))
#else
static void malloc_assert_failed(char *, const char *, int,
                                 const char *, const char *);
#  define xmalloc_assert(expr)  do {                                          \
          (expr) ? ((void)(0)) :                                              \
          malloc_assert_failed(__STRING(expr), file, line, func,              \
                               __func__);                             \
          } while (0)
#endif /* NDEBUG */
~~~~

~~~~
#ifndef NDEBUG
static void malloc_assert_failed(char *expr, const char *file,
		                 int line, const char *caller, const char *func)
{
	error("%s() Error: from %s:%d: %s(): Assertion (%s) failed",
	      func, file, line, caller, expr);
	abort();
}
#endif
~~~~